lib/features/interactive.js

```diff
@@ -0,0 +1,151 @@
+import assert from 'assert';
+import Debug from 'debug';
+import {DockerExecServer} from 'docker-exec-websocket-server';
+import fs from 'fs-ext';
+import http from 'http';
+import https from 'https';
+import path from 'path';
+import Promise from 'promise';
+import slugid from 'slugid';
+import SharedFileLock from '../shared_file_lock';
+import url from 'url';
+let debug = Debug('docker-worker:features:interactive');
+
+//Number of attempts to find a port before giving up
+const MAX_RANDOM_PORT_ATTEMPTS = 20;
+
+export default class WebsocketServer {
+  constructor () {
+    let id = slugid.v4();
+    this.path = '/' + id;
+    this.lock = path.join('/tmp/', id + '.lock');
+  }
+
+  async link(task) {
+    this.lockFile = await Promise.denodeify(fs.open)(this.lock, 'w');
+
+    return {
+      binds: [{
+        source: path.join(__dirname, '../../bin-utils'),
+        target: '/.taskcluster_utils',
+        readOnly: true
+      }, {
+        source: this.lock,
+        target: '/tmp/interactive.lock',
+        readOnly: false
+      }]
+    };
+  }
+
+  async started(task) {
+    debug('creating ws server');
+    let httpServ;
+    if (task.runtime.interactive.ssl) {
+      let readFile = Promise.denodeify(fs.readFile);
+      let [key, cert] = await Promise.all([
+        readFile(task.runtime.ssl.key),
+        readFile(task.runtime.ssl.certificate)
+      ]);
+      httpServ = https.createServer({key, cert});
+    } else {
+      httpServ = http.createServer();
+    }
+
+    let port;
+    //searching for an open port between 32768 and 61000
+    let attempts = 0;
+    while (true) {
+      port = Math.floor((Math.random() * (61000 - 32768)) + 32768); 
+      try {
+        await (new Promise((resolve, reject) => {
+          httpServ.listen(port);
+          httpServ.once('listening', () => {
+            httpServ.removeListener('error', reject);
+            resolve();
+          });
+          httpServ.once('error', reject)}));
```

Comment:
if we more than `12` iterations or so node will complain about event handler leaking... It's not really a problem, because promise won't resolve more than once. But if we don't remove the `error` handler  and we retry to we change the semantics... Not handling an event called `error` is a special case that crashes the process.

```js
await new Promise((resolve, reject) => {
  let listeningHandler, errorHandler;
  listeningHandler = () => {
    httpServ.removeListener('error', errorHandler);
    resolve();
  };
  errorHandler = (err) => {
    httpServ.removeListener('error', listeningHandler);
    reject(err);
  };
  httpServ.once('listening', listeningHandler);
  httpServ.once('error', listeningHandler);
  httpServ.listen(port);
});
```

Note, we might also want to listen for the `error` event from `httpServ` or not... I'm not sure what we would do if it happens... I suspect it doesn't really happen once we've started listening anyways.